DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/src/svd.hpp

```diff
@@ -0,0 +1,273 @@
+
+  // USM allocations
+  T *input_matrix_device;
+  ac_int<1, false> *rank_deficient_flag_device;
+  T *u_matrix_device;
+  T *s_matrix_device;
+  T *v_matrix_device;
+  T *r_device;
+
+  if (q.get_device().has(sycl::aspect::usm_device_allocations)) {
+    std::cout << "Using device allocations" << std::endl;
+    // Allocate FPGA DDR memory.
+    input_matrix_device = sycl::malloc_device<T>(kInputMatrixSize, q);
+    rank_deficient_flag_device = sycl::malloc_device<ac_int<1, false>>(1, q);
+    u_matrix_device = sycl::malloc_device<T>(kUMatrixSize, q);
+    s_matrix_device = sycl::malloc_device<T>(kInputMatrixSize, q);
+    v_matrix_device = sycl::malloc_device<T>(kEigenVectorsMatrixSize, q);
+    r_device = sycl::malloc_device<T>(kRMatrixSize, q);
+  } else if (q.get_device().has(sycl::aspect::usm_shared_allocations)) {
+    std::cout << "Using shared allocations" << std::endl;
+    // No device allocations means that we are probably in a SYCL HLS flow
+    input_matrix_device = sycl::malloc_shared<T>(kInputMatrixSize, q);
+    rank_deficient_flag_device = sycl::malloc_shared<ac_int<1, false>>(1, q);
+    u_matrix_device = sycl::malloc_shared<T>(kUMatrixSize, q);
+    s_matrix_device = sycl::malloc_shared<T>(kInputMatrixSize, q);
+    v_matrix_device = sycl::malloc_shared<T>(kEigenVectorsMatrixSize, q);
+    r_device = sycl::malloc_shared<T>(kRMatrixSize, q);
+  } else {
+    std::cerr << "USM device allocations or USM shared allocations must be "
+                 "supported to run this sample."
+              << std::endl;
+    std::terminate();
+  }
+
+  // Check that the malloc succeeded.
+  if (input_matrix_device == nullptr) {
+    std::cerr << "Error when allocating the input matrix." << std::endl;
+    std::terminate();
+  }
+  if (rank_deficient_flag_device == nullptr) {
+    std::cerr << "Error when allocating the rank deficient flag." << std::endl;
+    std::terminate();
+  }
+  if (u_matrix_device == nullptr) {
+    std::cerr << "Error when allocating the U matrix." << std::endl;
+    std::terminate();
+  }
+  if (s_matrix_device == nullptr) {
+    std::cerr << "Error when allocating the S matrix." << std::endl;
+    std::terminate();
+  }
+  if (v_matrix_device == nullptr) {
+    std::cerr << "Error when allocating the V matrix." << std::endl;
+    std::terminate();
+  }
+  if (r_device == nullptr) {
+    std::cerr << "Error when allocating the R matrix." << std::endl;
+    std::terminate();
+  }
+
+  // copy input matrix to device
+  q.memcpy(input_matrix_device, a_matrix.data(),
+           kInputMatrixSize * sizeof(float))
+      .wait();
+
+  // read input matrix from DDR
+  auto ddr_write_event = q.submit([&](sycl::handler &h) {
+    h.single_task<IDInputMatrixFromDDRToLocalMem>(
+        [=]() [[intel::kernel_args_restrict]] {
+          MatrixReadFromDDRToPipeByBlocks<
+              T, cols, rows, kNumElementsPerDDRBurst, InputMatrixPipe>(
+              input_matrix_device, 1, repetitions);
+        });
+  });
```

Comment:
```suggestion
  auto ddr_write_event = q.submit([&](sycl::handler &h) {
    h.single_task<IDInputMatrixFromDDRToLocalMem>(
        [=]() [[intel::kernel_args_restrict]] {
          MatrixReadFromDDRToPipeByBlocks<
              T, cols, rows, kNumElementsPerDDRBurst, InputMatrixPipe>(
              input_matrix_device, 1, repetitions);
        });
  });
```

It would be more accurate to a real-world workload to re-launch the kernels multiple times (even if it's always on the same data) than to have the repetition counter built into the device code. I'm concerned that reporting benchmarks generated in this way will set users up for unrealistic expectations.